text/0000-volta-run.md

```diff
@@ -0,0 +1,72 @@
+
+# Pedagogy
+[pedagogy]: #pedagogy
+
+We will need to document the existence of `volta run` and how the users can specify the versions of Node and Yarn when running a command. This will be provided on the command line by StructOpt, however it should also be explained in detail in the documentation. We should also document how the overall platform is resolved if the user specifies only one of Node or Yarn, so that the behavior is clear to users.
+
+# Details
+[details]: #details
+
+The new command will have the structure:
+
+```
+volta run [--node <version>] [--yarn <version>] <command>
+```
+
+Specifically, the `--node` and `--yarn` settings will need to be specified _before_ the command to be executed, as everything after the command will be treated as arguments to the command itself. This will allow users to specify a full command, with any command-line flags they need, without having to append `--` to separate the `volta run` command from the others.
+
+## Platform Resolution
+
+If the user specifies only one (or none) of the platform settings when calling `volta run`, the remaining settings will fall back to the existing context-aware logic. So if the user is in a project with `node@10.16.0` and `yarn@1.17.3` pinned, and runs:
+
+```
+volta run --node 8 ember serve
+```
+
+We will resolve the a platform that has `node@8` and `yarn@1.17.3`, with the latter pulled from the existing pinned project and the former overridden by the command-line flag. This allows users to specify the parts of the platform that they want to change, while avoiding the necessity of specifying everything, even those things that don't change.
+
+## 3rd-Party Binaries
+
+After resolving the platform, if the command the user is executing is a 3rd party binary (not `node`, `npm`, `npx`, or `yarn`), then we will still do resolution of that binary relative to the current project (if any). That means if we are inside a project with that binary as a dependency, we will still use the project-local version of that tool.
+
+## Verbose Output
+
+In order to provide meaningful verbose output, we will need to update the `SourcedPlatformSpec` model to support platforms that come from the command line as well. This will likely require a refactor to support specifying the version for each tool explicitly, as opposed to having a single `Source` for the Platform as a whole.
+
+# Critique
+[critique]: #critique
+
+## Command Name
+
+There are alternative names we could use for the command, such as:
+
+- `volta exec`
+- `volta do`
+
+However, `volta run` aligns with the "run script" command in both `npm` and `yarn`, so it seems to fit the best for this use-case.
+
+## Platform Resolution
+
+Instead of inheriting the platform from the current, we could treat it as purely defined on the command-line. This would give users total control over all of the pieces of the platform, including allowing them to remove `yarn`, even if it is specified in the project or default platforms.
```

Comment:
I'd prefer to do `--no-yarn` style option to support the "don't fallback" style.

Reply:
I like that better than the example I had in "Unresolved Questions" of `--yarn none`. Still open if we want to support that, but I can see it being useful.